### PR TITLE
feat: any-to-any onnx-genai pipeline support (VLM/ASR/codec/GGUF + Qwen3-TTS)

### PR DESCRIPTION
**Pull request by justinchuby**
## Summary

Adds **model-builder + metadata-emitter support for any-to-any onnx-genai pipelines**, pairing with the onnx-genai `feat/any-to-any-pipeline` PR (runtime). Mobius now builds and self-describes VLM, ASR, audio-codec, GGUF, and multi-decoder TTS packages for the onnx-genai composite runtime.

## What's included

**onnx-genai metadata emitters** (`src/mobius/integrations/onnx_genai/`)
- Composite **ASR** (Whisper-style), **audio-codec** (audio-to-audio), and **multimodal VLM** pipeline metadata.
- Self-contained packages: emit `tokenizer.json` for decoder/VLM/ASR builds.
- **`build_tts_pipeline_metadata`**: emits the `pre_embedder`-driven `nested_autoregressive` contract for real Qwen3-TTS (talker + code_predictor + talker_step_embedder), validated against onnx-genai's committed JSON schema. Packages without the pre-embedder raise a precise error.
- Structural dispatch in `write_onnx_genai_config` (fails loudly on unknown shapes).

**GGUF integration** (`src/mobius/integrations/gguf/`)
- `build-gguf --runtime onnx-genai` emits `inference_metadata` + `tokenizer.json`.
- Build **Gemma3/Gemma4 from GGUF**; **Gemma4 multimodal VLM from mmproj GGUF** with quantization preserved (~4.2 GB self-contained).
- Reconstruct `tokenizer.json` for archs transformers doesn't recognize (gemma4); fix string-array reader off-by-one.

**Qwen3-TTS model** (`src/mobius/models/qwen3_tts.py`, `tasks/_tts.py`)
- `talker_step_embedder` pre-embedding component (`frame_codes [+ text_embed] -> inputs_embeds`), sharing weights with embedding/code_predictor (validated maxdiff 0.0 vs HF; exact code match over full generation).
- Build fixes: route the Identity-folded `codec_embeddings` initializer; preserve the talker's non-standard interleaved-MRoPE `rope_scaling` (dropped by HF standardization) so the talker uses MRoPE.

## Testing

onnx-genai integration + qwen3-tts model tests green; TTS emitter validated end to end against the real 1.7B `Qwen/Qwen3-TTS-12Hz-1.7B-CustomVoice` package + the committed JSON schema.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: src/mobius/integrations/gguf/__init__.py [hunk 1]
```diff
@@ -16,5 +16,11 @@
...
 
 from mobius.integrations.gguf._builder import build_from_gguf
-
-__all__ = ["build_from_gguf"]
+from mobius.integrations.gguf._mmproj import build_gemma4_vlm_from_gguf
+from mobius.integrations.gguf._tokenizer import write_gguf_tokenizer_json
+
+__all__ = [
+    "build_from_gguf",
+    "build_gemma4_vlm_from_gguf",
```
Comment (justinchuby) [inline src/mobius/integrations/gguf/__init__.py:31]: Should this just keep build_from_gguf? Or maybe add a separate mmproj handler?

File: src/mobius/__main__.py [hunk 2]
```diff
@@ -667,6 +708,17 @@ def main(argv: list[str] | None = None) -> None:
             "Defaults to 'default' (portable ONNX, no vendor fusions)."
         ),
     )
+    gguf_parser.add_argument(
+        "--runtime",
+        choices=["onnx-genai"],
```
Comment (justinchuby) [inline src/mobius/__main__.py]: Add ort-genai too.

File: src/mobius/integrations/gguf/__init__.py (hunk 1 above)
Comment (justinchuby) [reply src/mobius/integrations/gguf/__init__.py:31]: Unified it: `build_from_gguf` now takes an optional `mmproj=` param and delegates to the multimodal VLM builder when set, so there's a single public entry point for both text-only and multimodal GGUF builds. The CLI (`build-gguf --mmproj`) uses it. `build_gemma4_vlm_from_gguf` stays available for direct use. Done in 49da57230c114d5a3de81746d81c596b44b39dbc.

File: src/mobius/__main__.py (hunk 2 above)
Comment (justinchuby) [reply src/mobius/__main__.py]: Added — `build-gguf --runtime` now offers `{ort-genai, onnx-genai}` (matching the main `build` command). The save path already handled `ort-genai`; only the argparse choices had omitted it. Done in 49da57230c114d5a3de81746d81c596b44b39dbc.

**Comment by justinchuby** [comment]
Addressed the review comments in `5a5e02f` (lintrunner clean, tests green). Summary:

**Second review round (fixed now):**
- **Prefill embedder batch mismatch** (`qwen3_tts.py`): the batch=1 codec/special pieces are now broadcast to the dynamic batch `B` before `Concat`, fixing the batch-dim failure for `B>1`; added `test_prefill_embedder_batched`.
- **Prefill embedder vocab validation**: `__init__` now asserts the fixed TTS special-token / codec-prefill ids fit `text_vocab_size` / `vocab_size`.
- **Hardcoded `dtype: "fp32"`** in multimodal/ASR/TTS dataflow edges: now derived from the model's activation dtype (`_activation_dtype_tag(config)` → fp16/bf16/fp32), so quantized/half builds are described correctly.
- **`_looks_like_multimodal`**: now also requires the `embedding` fusion component, so we never emit multimodal metadata referencing a missing embedding model.
- **mmproj VLM build bypassing optimizations**: `build_gemma4_vlm_from_gguf` now routes through `build_from_module` (EP-aware `optimize_model` passes), matching the text-only `build_from_gguf` path.
- **`gemma4._make_lm_head`**: guarded the `type | None` linear class.
- **`_config_resolver` rope restore**: raw nested `rope_scaling`/`rope_parameters` are now restored whenever HF standardization dropped **or rewrote** them (not only when `None`).

**First round (already fixed earlier in this PR):** empty-except → explicit return; `_audio_codec_codes_dtype` FLOAT16/BFLOAT16 mapping; `onnx_ir.save` instead of `onnx.save(ir.to_proto())`; `TTSTask`/`_looks_like_multi_decoder_tts` docstrings; speech-to-text + TTS builders/writers exported from `mobius.integrations.onnx_genai` (+ `__all__`); unified `build_from_gguf(mmproj=...)`; `build-gguf --runtime {ort-genai,onnx-genai}`.

File: src/mobius/models/qwen3_tts.py [hunk 3]
```diff
@@ -590,11 +862,20 @@ def finalize_stacked_weights(
...
+        # exact initializer name so it is applied (apply_weights routes an unprefixed
+        # name to the only component that declares it).
         if codec_emb_weights:
             num_groups = max(codec_emb_weights.keys()) + 1
             stacked = torch.stack([codec_emb_weights[i] for i in range(num_groups)])
-            remaining["code_predictor.stacked_codec_embedding"] = stacked
+            remaining["codec_embeddings"] = stacked
+            # Share the same stacked table with the talker step embedder so it
+            # can materialize codec_sum in-graph (Σ cp_codec_weights[i][code]).
+            remaining["talker_step_embedder.stacked_codec_embedding"] = stacked
```
[the remark this answers is not in the document]
Comment (justinchuby) [reply src/mobius/models/qwen3_tts.py:891]: Reasonable — done in 4ae2e0b. The stacked codec table is now keyed under **both** `codec_embeddings` (Identity-folded case) and `code_predictor.stacked_codec_embedding` (unfolded case). `apply_weights` populates whichever initializer the build actually emits and only logs the other as unmapped (it doesn't error on extra keys), so the table is applied robustly without depending on the IR fold.
